[RzIL] Add lifter for 32-bit Arm, including thumb

thestr4ng3r opened the pull request:
# SQUASH THIS

XVilka commented on librz/il/il_export.c:
```diff
@@ -244,11 +244,11 @@ static void il_opdmp_is_zero(RzILOpPure *op, RzStrBuf *sb, PJ *pj) {
 
 static void il_opdmp_lognot(RzILOpPure *op, RzStrBuf *sb, PJ *pj) {
-	il_op_param_1("lognot", op->op.lognot, bv);
+	il_op_param_1("~", op->op.lognot, bv);
```
Please extract these changes in a separate PR

thestr4ng3r replied:
Done: https://github.com/rizinorg/rizin/pull/2285

thestr4ng3r committed:
WIP Arm IL Lifter

XVilka commented on librz/analysis/arch/arm/arm_il32.c:
```diff
@@ -0,0 +1,2695 @@
+}
+
+#define PC(addr, is_thumb)      (addr + (is_thumb ? 4 : 8))
+#define PCALIGN(addr, is_thumb) (PC(addr, is_thumb) & ~3)
```
Shouldn't `3` have a corresponding number suffix? `UL` or something like that?

thestr4ng3r replied:
Sure doesn't hurt at least.

XVilka commented on librz/analysis/arch/arm/arm_il32.c:
```diff
@@ -0,0 +1,2695 @@
+		arm_reg qreg = ARM_REG_Q0 + idx / 2;
+		ut64 mask[2] = { UT64_MAX, UT64_MAX };
+		mask[idx % 2] = 0;
+		RzILOpBitVector *masked = LOGAND(read_reg(0, qreg), rz_il_op_new_bitv(rz_bv_new_from_bytes_le((const ut8 *)mask, 0, 128)));
```
What is 128 here?

thestr4ng3r replied:
Size of a q reg, but I am rewriting this anyway.

XVilka commented on librz/analysis/arch/arm/arm_il32.c:
```diff
@@ -0,0 +1,2695 @@
+		return NULL;
+	}
+	return SEQ2(
+		SETG("lr", U32(((insn->address + insn->size) & ~1) | (is_thumb ? 1 : 0))),
```
Same, shouldn't it be `~1UL`?

XVilka commented on librz/analysis/arch/arm/arm_il32.c:
```diff
@@ -0,0 +1,2695 @@
+	}
+	// There are more bits in ARM, but this is all we have:
+	return write_reg(REGID(0),
+		LOGOR(ITE(VARG("nf"), U32(1 << 31), U32(0)),
```
Nice stair is right here!

thestr4ng3r replied:
You can thank clang-format

XVilka commented on librz/analysis/arch/arm/arm_il32.c:
```diff
@@ -0,0 +1,2695 @@
+		return NULL;
+	}
+	if (set_ge) {
+		eff = SEQ2(
```
Is it possible to simplify this somehow?

thestr4ng3r replied:
Actually yes it can, good point.

XVilka reviewed (changes requested):
Overall looks good. Amazing job! It's a pity capstone isn't being very helpful for providing necessary opcode information. Evidently, It would greatly simplify some parts of the code.

thestr4ng3r committed:
Address review

thestr4ng3r committed:
Use dregs instead of qregs as base

thestr4ng3r committed:
Add vmov and fix sregs

thestr4ng3r committed:
Fix ldm overriding base

thestr4ng3r committed:
Fix asrs carry

XVilka reviewed (approved):
LGTM

thestr4ng3r committed:
Fix rrxs carry

thestr4ng3r committed:
Patch a capstone bug

thestr4ng3r committed:
Fix rbit

thestr4ng3r committed:
Fix shadd16

thestr4ng3r committed:
Fix smlabb cf

thestr4ng3r committed:
Fix a lot of smla

thestr4ng3r committed:
Fix leaks in RzIL SHIFTRA macro usage

The v argument may have been evaluated twice and thus leak depending on
how the macro was used.

thestr4ng3r committed:
Fix sdiv leak

thestr4ng3r committed:
Fix ldr leak

thestr4ng3r committed:
Fix ssat16 qf

wargio reviewed (commented):
lgtm. <3
